### ISSUE TITLE
`find_essential` backward raises `_LinAlgError` in the `linalg.eigvals` backward on degenerate samples

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4831]
On `main` (`012997ceb`, macOS arm64 CPU, torch 2.14.0), `find_essential` runs forward on some degenerate correspondence sets but raises in `.backward()`:

```python
import torch
from kornia.geometry.epipolar import find_essential

g = torch.Generator().manual_seed(19)
p = torch.rand(1, 5, 2, generator=g, dtype=torch.float64).requires_grad_()
E = find_essential(p, p, torch.ones(1, 5, dtype=torch.float64))  # forward returns
E.nan_to_num().sum().backward()
# torch._C._LinAlgError: torch.linalg.solve: (Batch element 0): The solver failed because the input matrix is singular.
```

8 of 100 random `p1 == p2` samples (seeds 0–99, float64) raise like this. The raising op is the backward of `torch.linalg.eigvals(C)` in `_null_to_Nister_solution_script`. With the companion matrix detached (`eigvals(C.detach())`), the same backward succeeds. When `C` is defective (repeated eigenvalues), the eigenvector matrix is singular, and the eig backward solves against it. This happens even though the upstream gradient for those candidates is zero, for example when they are NaN-masked or replaced by the identity fallback. One such sample makes the whole batch's backward fail.

After #4766, an exactly singular sample (e.g. all points at the origin) returns the identity fallback in the forward pass but hits the same error in the backward pass.

A possible direction: for elements whose candidates are discarded anyway, replace `C` with a matrix with distinct eigenvalues before `eigvals`, so the backward stays finite and their gradient is zero. This does not help a regular but defective `C`, which needs a decision on what gradient to return.

Related: #4820 (forward abort in `eigvals` on a non-finite companion matrix).

Posted on behalf of @ducha-aiki by Claude (Opus 5.5).

**Comment by VanniLeonardo** [comment]
I will take this one too, right after #4820, since both fail at the same `eigvals` call. A maintainer will need to assign me.

**Comment by VanniLeonardo** [comment]
I reproduced this on Linux x86 with torch 2.9.1 and traced it. Before I write code there is one decision that is yours, so here is what I measured.

**Reproduction.** On `main` `9c39a109`, the backward of `find_essential(p, p, w).nan_to_num().sum()` raises for 7 of 100 random `p1 == p2` samples at float64 and 10 of 100 at float32, with `p = torch.rand(1, 5, 2)` for seeds 0 to 99. It never raises for 100 random pairs, and it always raises for points at the origin. The seeds differ from yours, since the eigen decomposition is platform-dependent.

**None of these are discarded samples.** `A10` is not singular in any of them, and they keep 1 to 10 finite candidates. The companion matrix has an exactly repeated eigenvalue at 0, of multiplicity 3 to 9, and its eigenvector matrix is singular, so the `eigvals` backward cannot solve against it. Of the 96 candidates built from that repeated root, 82 are returned as finite output. A placeholder for discarded elements alone would only fix the origin input.

**Two causes, and one is #4847.** In the failing samples the leading coefficient of the degree-10 polynomial is either exactly zero (5 of 7 at float64, 7 of 10 at float32) or negative (2 of 7 and 3 of 10). A negative one is clamped to `+1e-8` by the regression in #4847. With #4848, which fixes that, and #4846 on top, the failures drop to 5 of 100 at float64 and 8 of 100 at float32. At float64 all 5 have a zero leading coefficient, and at float32 7 of the 8 do.

**The zero leading coefficient is the real case here.** It means the polynomial has a lower degree. The `1e-8` floor then gives a companion row of about 1e35, and LAPACK returns a repeated zero eigenvalue. On `p1 == p2` data this is common: 50 of 100 samples at float64 and 45 of 100 at float32 have it, not only the ones that raise. Their finite candidates satisfy the epipolar constraint to about 1e-16, but none is an essential matrix. The best residual of `det E = 0` and `2 E E^T E - tr(E E^T) E = 0` at unit norm is between 3e-6 and 0.15, and typically 1e-3 to 1e-2.

**What I tested.** On top of #4846 and #4848, I routed singular elements and elements with an exactly zero leading coefficient to the identity placeholder from #4846, and set their candidates to NaN, so they take the identity fallback. Backward failures go from 5 to 0 of 100 at float64 and from 8 to 1 at float32. The outputs and input gradients of 120 random pairs stay bitwise identical. Two things remain.

- One float32 sample with a negative leading coefficient still raises. I have not traced it yet.
- For points at the origin the backward no longer raises in `eigvals`, but the gradient is NaN. Anomaly detection puts that in `LinalgSvdBackward0`, the SVD of the rank-1 design matrix, which has repeated singular values. No placeholder at `eigvals` can change that.

**The decision.** What should a sample with an exactly zero leading coefficient return?

1. **No usable roots, identity fallback**, as tested above. This is the smallest change and consistent with #4766 and #4846. It changes the forward pass for those degenerate samples, but only by dropping candidates that fail the essential-matrix constraints.
2. **Deflate the polynomial** to its true degree and return its roots, with the rest set to NaN. This keeps any valid roots but is more code, and I have not measured it yet.
3. **Keep the forward pass** and only stop the backward from raising, by treating the roots of a defective companion matrix as constants. The gradient is then finite but not the true derivative, which is unbounded at a multiple root.

I would go with 1. Is that acceptable? And do you want the NaN from the SVD backward on a rank-1 design matrix handled here, or as a separate issue? This builds on #4846 and #4848, so I would open the PR once those land.

**Comment by ducha-aiki** [comment]
@VanniLeonardo it's yours. We can't assign outside contributors, so this comment is the claim. #4846 and #4848 have both merged, so you can build on `main` now. I'll answer your option question (zero leading coefficient: identity fallback, deflation, or constant roots) and the rank-1 SVD question in a separate comment here.

Posted on behalf of @ducha-aiki by Claude (Opus 5.5).

**Comment by ducha-aiki** [comment]
@VanniLeonardo go with **option 1**. A sample whose degree-10 polynomial has an exactly zero leading coefficient has no usable roots. Route it to the identity placeholder from #4846 and set its candidates to NaN, so it takes the same fallback as #4766 and #4846. Deflation (option 2) is more code for candidates that, by your measurement, don't satisfy the essential-matrix constraints anyway. Option 3 returns a gradient that is finite but wrong, which is worse than the current raise.

On scope:

- **The origin input, where the NaN comes from `LinalgSvdBackward0` on the rank-1 design matrix, stays out of this PR.** It is the SVD backward, not `eigvals`, so it is closer to #4855 and your #4860. Check it against #4860 once that is ready. If #4860 doesn't cover it, open a separate issue with the anomaly trace.
- **The one float32 sample with a negative leading coefficient that still raises:** trace it before opening the PR. Either fix it here if it is the same mechanism, or report it in the PR description as a separate case with its seed.
- **Changelog:** option 1 changes the forward output for those samples, since finite candidates that used to be returned become the fallback. So use a `.breaking` fragment, as #4848 did, and say what was returned before.
- **Tests:** pin the backward no longer raising, and pin the forward output of ordinary pairs staying bitwise unchanged. Include a mutant check that the zero-coefficient guard is actually what the new pin exercises.

Posted on behalf of @ducha-aiki by Claude (Opus 5.5).

**Comment by ducha-aiki** [comment]
Platform data from the batch-8a CI (#4887): the backward raise on an all-zero 5-point sample happens on macOS arm64 with torch 2.14.0 and 2.5.1, and on GitHub's ubuntu x86 runner with torch 2.14.0 float32, but **not** on ubuntu with torch 2.5.1 float64 — there `find_essential(...).sum().backward()` returns. Whether `linalg.eigvals`' backward raises on the degenerate companion matrix depends on the LAPACK build, so #4887 documents this issue in the `find_essential` block without a wart pin; a fix that never reaches the raising path (a validity check before `eigvals`, or #4855's null-space treatment) is testable on every platform.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).
